### PR TITLE
ARROW-11776: [C++][Java] Support parquet write from ArrowReader to file

### PR DESCRIPTION
**Pull request by JkSelf**
This PR is aim to support parquet write from ArrowReader to file.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
@JkSelf Can we just use the existing ticket ARROW-11776?

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/NativeScannerAdaptorImpl.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+         (byte) ((data >> 0) & 0xff),
+         (byte) ((data >> 8) & 0xff),
+         (byte) ((data >> 16) & 0xff),
+         (byte) ((data >> 24) & 0xff),
+         (byte) ((data >> 32) & 0xff),
+         (byte) ((data >> 40) & 0xff),
+         (byte) ((data >> 48) & 0xff),
+         (byte) ((data >> 56) & 0xff),
+     };
+    }
```
Comment (zhztheplayer) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/NativeScannerAdaptorImpl.java:133]: Is it necessary to conduct this kind of encoding?

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.arrow.dataset.scanner.Scanner;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.AutoCloseables;
+import org.apache.arrow.vector.util.SchemaUtility;
+
+/**
+ * JNI-based utility to write datasets into files. It internally depends on C++ static method
+ * FileSystemDataset::Write.
+ */
+public class DatasetFileWriter {
```
Comment (zhztheplayer) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java:30]: Let's add some unit tests.

You can refer to the previous PR https://github.com/apache/arrow/pull/10201/files

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/NativeScannerAdaptorImpl.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import org.apache.arrow.dataset.scanner.ScanTask;
+import org.apache.arrow.dataset.scanner.Scanner;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.ipc.ArrowReader;
+
+/**
+ * Default implementation of {@link NativeScannerAdaptor}.
+ */
+public class NativeScannerAdaptorImpl implements NativeScannerAdaptor, AutoCloseable {
```
Comment (zhztheplayer) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/NativeScannerAdaptorImpl.java:35]: Is it needed to create this class?

`IteratorImpl` seems to already handle everything.

File: java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeRecordBatchIterator.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package org.apache.arrow.dataset.jni;
+
+import java.util.Iterator;
+/**
+ * Iterate on the memory address of ArrowArray
+ * next() should be called from C++ scanner to read the memory address of ArrowArray.
+ */
+public interface NativeRecordBatchIterator extends Iterator<byte[]>, AutoCloseable {
```
Comment (zhztheplayer) [inline java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeRecordBatchIterator.java:25]: May be we can rename this as `CRecordBatchIterator` or something to make things more clearer.

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 5]
```diff
@@ -170,6 +177,116 @@ class DisposableScannerAdaptor {
...
+};
+
+/// \brief Create scanner that scans over Java dataset API's components.
+///
+/// Currently, we use a NativeRecordBatchIterator as the underlying
+/// Java object to do scanning. Which means, only one single task will
+/// be produced from C++ code.
+arrow::Result<std::shared_ptr<arrow::dataset::Scanner>> MakeJavaDatasetScanner(
+    JavaVM* vm, jobject iter, std::shared_ptr<arrow::Schema> schema) {
+  arrow::RecordBatchIterator itr = arrow::MakeFunctionIterator(
```
Comment (zhztheplayer) [inline java/dataset/src/main/cpp/jni_wrapper.cc:257]: There might be some better solution to distinguish on the 2 variables than just using `iter` and `itr`?

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/NativeScannerAdaptorImpl.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+         (byte) ((data >> 32) & 0xff),
+         (byte) ((data >> 40) & 0xff),
+         (byte) ((data >> 48) & 0xff),
+         (byte) ((data >> 56) & 0xff),
+     };
+    }
+
+    @Override
+    public byte[] next() {
+      ArrowArray arrowArray = ArrowArray.allocateNew(allocator);
```
Comment (zhztheplayer) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/NativeScannerAdaptorImpl.java:137]: `arrowArray` may need to be eventually closed or memory may leak

**Comment by lidavidm** [comment]
Why isn't this just using the C Data Interface instead of doing things like serializing schemas and manually adapting iterators?

**Comment by lidavidm** [comment]
In particular you can export a stream now and that will arrive in C++ as a RecordBatchReader

**Comment by JkSelf** [comment]
@zhztheplayer  @lidavidm 
Sorry for the delay response. I have resolved the comments. Please help to review again. Thanks.

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java [hunk 7]
```diff
@@ -45,4 +46,21 @@ private JniWrapper() {
...
+   * @param itr iterator to be used for writing
+   * @param schema serialized schema of output files
+   * @param fileFormat target file format (ID)
+   * @param uri target file uri
+   * @param partitionColumns columns used to partition output files
+   * @param maxPartitions maximum partitions to be included in written files
+   * @param baseNameTemplate file name template used to make partitions. E.g. "dat_{i}", i is current partition
+   *                         ID around all written files.
+   */
+  public native void writeFromScannerToFile(CRecordBatchIterator itr, long schema_address,
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java:62]: Same question - why are we manually bridging data to C++, why do we have CRecordBatchIterator at all, when we could use ArrowArrayStream instead? This should just take an ArrowArrayStream as the data source

**Comment by lidavidm** [comment]
CC @lwhite1 @davisusanibar

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java (hunk 7 above)
Comment (JkSelf) [reply java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java]: We have already used the ArrowArrayStream sharing data to native. Here the CRecordBatchIterator is to iterate the ArrowArrayStream object. I have changed the CRecordBatchIterator name to CArrowArrayStreamIterator. Sorry for the wrong naming.

Comment (lidavidm) [reply java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java]: No, what I mean is that you do not need the iterator at all. Implement the ArrowReader API for Scanner or whatever the Java-side data source is, then export that _once_ and import it _once_ on the C++ side. The JNI side should not need to know about any new Java-side classes or methods.

Comment (JkSelf) [reply java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java]: Thanks for your detailed explanation. Already removed the iterator and use the ArrowArrayStream as the data source .

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java [hunk 8]
```diff
@@ -45,4 +45,20 @@ private JniWrapper() {
...
+   *
+   * @param schema_address the schema address
+   * @param fileFormat target file format (ID)
+   * @param uri target file uri
+   * @param partitionColumns columns used to partition output files
+   * @param maxPartitions maximum partitions to be included in written files
+   * @param baseNameTemplate file name template used to make partitions. E.g. "dat_{i}", i is current partition
+   *                         ID around all written files.
+   */
+  public native void writeFromScannerToFile(long stream_address, long schema_address,
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java:60]: ArrowArrayStream has a schema already

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      Iterator<? extends ScanTask> taskIterators = scanner.scan().iterator();
+      while (taskIterators.hasNext()) {
+        ArrowReader currentReader = taskIterators.next().execute();
+        ArrowArrayStream stream = ArrowArrayStream.allocateNew(allocator);
+        Data.exportArrayStream(allocator, currentReader, stream);
+        JniWrapper.get().writeFromScannerToFile(stream.memoryAddress(), arrowSchema.memoryAddress(),
+            format.id(), uri, partitionColumns, maxPartitions, baseNameTemplate);
+        currentReader.close();
+        stream.close();
+      }
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java:64]: I don't think this is right. We should implement ArrowReader over Scanner, then write the entire dataset in one go. And then this method can be used to write _any_ ArrowReader, not just Scanners.

**Review by lidavidm — CHANGES_REQUESTED** [review]

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java (hunk 9 above)
Comment (JkSelf) [reply java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java]: Added ArrowScannerReader over Scanner.

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 10]
```diff
@@ -170,6 +174,19 @@ class DisposableScannerAdaptor {
...
 
+std::shared_ptr<arrow::Schema> SchemaFromColumnNames(
+    const std::shared_ptr<arrow::Schema>& input,
+    const std::vector<std::string>& column_names) {
+  std::vector<std::shared_ptr<arrow::Field>> columns;
+  for (arrow::FieldRef ref : column_names) {
+    auto maybe_field = ref.GetOne(*input);
+    if (maybe_field.ok()) {
+      columns.push_back(std::move(maybe_field).ValueOrDie());
+    }
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_wrapper.cc:185]: We should error if a column name isn't present?

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 11]
```diff
@@ -510,3 +526,55 @@ Java_org_apache_arrow_dataset_file_JniWrapper_makeFileSystemDatasetFactory(
...
+  }
+
+  auto* arrow_stream = reinterpret_cast<ArrowArrayStream*>(c_arrow_array_stream_address);
+  struct ArrowSchema c_schema;
+  arrow_stream->get_schema(arrow_stream, &c_schema);
+
+  std::shared_ptr<arrow::RecordBatchReader> reader =
+      JniGetOrThrow(arrow::ImportRecordBatchReader(arrow_stream));
+  // Release the ArrowArrayStream
+  ArrowArrayStreamRelease(arrow_stream);
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_wrapper.cc:554]: ImportRecordBatchReader calls ArrowArrayStreamMove, so this is not necessary

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 12]
```diff
@@ -510,3 +526,55 @@ Java_org_apache_arrow_dataset_file_JniWrapper_makeFileSystemDatasetFactory(
...
+    jint max_partitions, jstring base_name_template) {
+  JNI_METHOD_START
+  JavaVM* vm;
+  if (env->GetJavaVM(&vm) != JNI_OK) {
+    JniThrow("Unable to get JavaVM instance");
+  }
+
+  auto* arrow_stream = reinterpret_cast<ArrowArrayStream*>(c_arrow_array_stream_address);
+  struct ArrowSchema c_schema;
+  arrow_stream->get_schema(arrow_stream, &c_schema);
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_wrapper.cc:549]: Once you import the reader, you can use `reader->schema()`.

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+   * @param maxPartitions maximum partitions to be included in written files
+   * @param partitionColumns columns used to partition output files. Empty to disable partitioning
+   * @param baseNameTemplate file name template used to make partitions. E.g. "dat_{i}", i is current partition
+   *                         ID around all written files.
+   */
+  public static void write(BufferAllocator allocator, ArrowReader reader, FileFormat format, String uri,
+                           String[] partitionColumns, int maxPartitions, String baseNameTemplate) {
+    RuntimeException throwableWrapper = null;
+    try {
+      ArrowArrayStream stream = ArrowArrayStream.allocateNew(allocator);
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java:54]: use try-with-resources

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+      Data.exportArrayStream(allocator, reader, stream);
+      JniWrapper.get().writeFromScannerToFile(stream.memoryAddress(),
+          format.id(), uri, partitionColumns, maxPartitions, baseNameTemplate);
+      stream.close();
+    } catch (Throwable t) {
+      throwableWrapper = new RuntimeException(t);
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java:60]: nit: why are we declaring the exception separately?

**Review by lidavidm — CHANGES_REQUESTED** [review]

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 15]
```diff
@@ -170,6 +172,21 @@ class DisposableScannerAdaptor {
...
+arrow::Result<std::shared_ptr<arrow::Schema>> SchemaFromColumnNames(
+    const std::shared_ptr<arrow::Schema>& input,
+    const std::vector<std::string>& column_names) {
+  std::vector<std::shared_ptr<arrow::Field>> columns;
+  for (arrow::FieldRef ref : column_names) {
+    auto maybe_field = ref.GetOne(*input);
+    if (maybe_field.ok()) {
+      columns.push_back(std::move(maybe_field).ValueOrDie());
+    } else {
+      return arrow::Status::Invalid("The provided column name is not in arrow schema");
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_wrapper.cc:184]:
```suggestion
      return arrow::Status::Invalid("Partition column '", ref.ToString(), "' is not in dataset schema");
```

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 16]
```diff
@@ -510,3 +526,49 @@ Java_org_apache_arrow_dataset_file_JniWrapper_makeFileSystemDatasetFactory(
...
+    JniThrow("Unable to get JavaVM instance");
+  }
+
+  auto* arrow_stream = reinterpret_cast<ArrowArrayStream*>(c_arrow_array_stream_address);
+  std::shared_ptr<arrow::RecordBatchReader> reader =
+      JniGetOrThrow(arrow::ImportRecordBatchReader(arrow_stream));
+  std::shared_ptr<arrow::dataset::ScannerBuilder> scanner_builder =
+      arrow::dataset::ScannerBuilder::FromRecordBatchReader(reader);
+  scanner_builder->Pool(arrow::default_memory_pool());
+  auto scanner = scanner_builder->Finish().ValueOrDie();
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_wrapper.cc:553]: We must check for error here

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 17]
```diff
@@ -510,3 +526,49 @@ Java_org_apache_arrow_dataset_file_JniWrapper_makeFileSystemDatasetFactory(
...
+  if (env->GetJavaVM(&vm) != JNI_OK) {
+    JniThrow("Unable to get JavaVM instance");
+  }
+
+  auto* arrow_stream = reinterpret_cast<ArrowArrayStream*>(c_arrow_array_stream_address);
+  std::shared_ptr<arrow::RecordBatchReader> reader =
+      JniGetOrThrow(arrow::ImportRecordBatchReader(arrow_stream));
+  std::shared_ptr<arrow::dataset::ScannerBuilder> scanner_builder =
+      arrow::dataset::ScannerBuilder::FromRecordBatchReader(reader);
+  scanner_builder->Pool(arrow::default_memory_pool());
```
Comment (lidavidm) [inline java/dataset/src/main/cpp/jni_wrapper.cc:552]: We must check for error here

File: java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Iterator;
+
+/**
+ * JNI-based utility to write datasets into files. It internally depends on C++ static method
+ * FileSystemDataset::Write.
+ */
+public class DatasetFileWriter {
+
+  /**
+   * Write a ArrowReader accepting Scanner into files.
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java:40]: The docstring is still incorrect

Comment (lidavidm) [reply java/dataset/src/main/java/org/apache/arrow/dataset/file/DatasetFileWriter.java:40]:
```suggestion
   * Write the contents of an ArrowReader as a dataset.
```

File: java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  @Override
+  public long bytesRead() {
+    return 0L;
+  }
+
+  @Override
+  protected void closeReadSource() throws IOException {
+    // no-op
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:99]: We need to close resources here

**Review by lidavidm — CHANGES_REQUESTED** [review]

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+      actualLoader.load(batch);
+    }
+
+    for (int i = 0; i < expectVsr.getFieldVectors().size(); i++) {
+      FieldVector vector = expectVsr.getFieldVectors().get(i);
+      FieldVector otherVector = actualVsr.getFieldVectors().get(i);
+      Assert.assertTrue(VectorEqualsVisitor.vectorEquals(vector, otherVector));
+    }
+
+    // Assert.assertTrue(expectVsr.equals(actualVsr));
```
Comment (lidavidm) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:137]: Commented code?

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    AutoCloseables.close(expectedBatches, actualBatches);
```
Comment (lidavidm) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:138]: Can we use try-finally for this?

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+    VectorSchemaRoot expectVsr = VectorSchemaRoot.create(expectedFactory.inspect(), rootAllocator());
+    VectorLoader expectLoader = new VectorLoader(expectVsr);
+    for(ArrowRecordBatch batch: expectedBatches) {
+      expectLoader.load(batch);
+    }
+
+    VectorSchemaRoot actualVsr = VectorSchemaRoot.create(actualFactory.inspect(), rootAllocator());
+    VectorLoader actualLoader = new VectorLoader(actualVsr);
+    for(ArrowRecordBatch batch: actualBatches) {
+      actualLoader.load(batch);
```
Comment (lidavidm) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:128]: This needs to be interleaved with the comparison, else you are actually only comparing the last batch

**Review by lidavidm — CHANGES_REQUESTED** [review]

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java (hunk 22 above)
Comment (JkSelf) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java]: Update to compare all the batches.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java (hunk 20 above)
Comment (JkSelf) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java]: removed.

File: java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+
+  @Override
+  public long bytesRead() {
+    return 0L;
+  }
+
+  @Override
+  protected void closeReadSource() throws IOException {
+    try {
+      currentTask.close();
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:100]: Is there an issue if we potentially double-close these?

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+      Assert.assertEquals(expectedBatches.size(), actualBatches.size());
+      VectorLoader expectLoader = new VectorLoader(expectVsr);
+      VectorLoader actualLoader = new VectorLoader(actualVsr);
+      for (int i = 0; i < expectedBatches.size(); i++) {
+        expectLoader.load(expectedBatches.get(i));
+        actualLoader.load(actualBatches.get(i));
+        for (int j = 0; j < expectVsr.getFieldVectors().size(); j++) {
+          FieldVector vector = expectVsr.getFieldVectors().get(i);
+          FieldVector otherVector = actualVsr.getFieldVectors().get(i);
+          Assert.assertTrue(VectorEqualsVisitor.vectorEquals(vector, otherVector));
```
Comment (lidavidm) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:132]: VectorSchemaRoot#equals should do this for you

Comment (JkSelf) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:132]: I try to use `VectorSchemaRoot#equals` to compare, but the result return false by [this issue](https://issues.apache.org/jira/browse/ARROW-18140) when comparing the schema info.

File: java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java (hunk 23 above)
Comment (JkSelf) [reply java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:100]: If already closed, the close method will directly return and no further operator.

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java (hunk 24 above)
Comment (lidavidm) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:132]: Thanks for filing that. Can you add a comment like `// TODO(ARROW-18140): ...`?

**Comment by lidavidm** [comment]
There are lint errors https://github.com/apache/arrow/actions/runs/3290191393/jobs/5426029066#step:6:7990

```

[WARN] /arrow/java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:28: Wrong order for 'java.io.IOException' import. [ImportOrder]
[WARN] /arrow/java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:31: Missing a Javadoc comment. [JavadocType]
[WARN] /arrow/java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:39:3: Missing a Javadoc comment. [JavadocMethod]
[WARN] /arrow/java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:61: 'if' construct must use '{}'s. [NeedBraces]
[WARN] /arrow/java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:86:5: WhitespaceAround: 'try' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] /arrow/java/dataset/src/main/java/org/apache/arrow/dataset/file/JniWrapper.java:60:50: Parameter name 'stream_address' must match pattern '^[a-z][a-zA-Z0-9]*$'. [ParameterName]
Error:  /arrow/java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:20:8: Unused import: java.io.ByteArrayOutputStream. [UnusedImports]
Error:  /arrow/java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:22:8: Unused import: java.io.IOException. [UnusedImports]
Error:  /arrow/java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:23:8: Unused import: java.nio.channels.Channels. [UnusedImports]
Error:  /arrow/java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:38:8: Unused import: org.apache.arrow.flatbuf.RecordBatch. [UnusedImports]
Error:  /arrow/java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:44:8: Unused import: org.apache.arrow.vector.ipc.WriteChannel. [UnusedImports]
Error:  /arrow/java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:46:8: Unused import: org.apache.arrow.vector.ipc.message.MessageSerializer. [UnusedImports]
```

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java (hunk 24 above)
Comment (JkSelf) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java:129]: Already added the comment.

**Comment by JkSelf** [comment]
@lidavidm 
Do you have any further comment?

**Comment by JkSelf** [comment]
> Can you rebase to see if the CI issue is fixed?

Rebased.

**Comment by lidavidm** [comment]
There's still a lint error here

```
Warning:  src/main/java/org/apache/arrow/dataset/scanner/ArrowScannerReader.java:[43,3] (javadoc) JavadocMethod: Missing a Javadoc comment.
```

**Review by lidavidm — APPROVED** [review]
Thank you!

**Pull request merged**